test/helper/sending.bash

```diff
@@ -82,20 +155,50 @@ function _send_email() {
 # chosen. Sending more than one mail at any given point in time with this function
 # is UNDEFINED BEHAVIOR!
 function _send_email_and_get_id() {
-  [[ -v CONTAINER_NAME ]] || return 1
+  # Get the name of the variable that the ID is stored in
+  local ID_ENV_VAR_NAME="${1:?Mail ID must be set for _send_email_and_get_id}"
+  # Get a "reference" to the content of ID_ENV_VAR_NAME so we can manipulate the content
+  local -n ID_ENV_VAR_REF=${ID_ENV_VAR_NAME}
+  # Export the variable so everyone has access
+  # shellcheck disable=SC2163
+  export "${ID_ENV_VAR_NAME}"
+  shift 1
 
-  _wait_for_empty_mail_queue_in_container
-  _send_email "${@}"
-  _wait_for_empty_mail_queue_in_container
-
-  local MAIL_ID
+  local QUEUE_ID MESSAGE_ID
   # The unique ID Postfix (and other services) use may be different in length
   # on different systems (e.g. amd64 (11) vs aarch64 (10)). Hence, we use a
   # range to safely capture it.
-  MAIL_ID=$(_exec_in_container tac /var/log/mail.log              \
-    | grep -E -m 1 'postfix/smtpd.*: [A-Z0-9]+: client=localhost' \
-    | grep -E -o '[A-Z0-9]{9,12}' || true)
+  local QUEUE_ID_REGEX='[A-Z0-9]{9,12}'
+  local MESSAGE_ID_REGEX="[0-9]{14}\\.${QUEUE_ID_REGEX}"
+
+  _wait_for_empty_mail_queue_in_container
+  if [[ ${1} == --expect-rejection ]]; then
+    shift 1
+    local OUTPUT=$(_send_email_unchecked "${@}")
+    # Because we expect the mail to be rejected, we have to query the mail log
+    # instead of `swaks`, because `swaks` cannot provide us with a queue ID when
+    # mail is rejected (we see something like this instead: `<** 554 5.7.1 Gtube pattern`).
+    QUEUE_ID=$(_exec_in_container tac /var/log/mail.log       \
+      | grep -E "postfix/smtpd.*: ${QUEUE_ID_REGEX}: client=" \
+      | grep -E -m 1 -o '[A-Z0-9]{9,12}' || :)
+  else
+    local OUTPUT=$(_send_email "${@}")
+    # When mail is expected to be delivered, we can use the output of `swaks`
+    # to easily query the queue ID.
+    QUEUE_ID=$(grep -F 'queued as' <<< "${OUTPUT}" | grep -E -o "${QUEUE_ID_REGEX}$")
+  fi
```

Comment:
Okay, I see. This simplifies `_send_email_and_get_id` even further, which I like. I will apply the changes manually because I see some gotchas still in the proposal.